packages/common/index.ts

```diff
@@ -1463,6 +1477,376 @@ export class CourseSettingsRequestBody {
+      (tag: any) => tag.display_name,
+    )
+    if (tagDisplayNames.length !== new Set(tagDisplayNames).size) {
+      const duplicateDisplayName = findFirstDuplicate(tagDisplayNames)
+      return `Tag display names must be unique. Duplicate display name found: "${duplicateDisplayName}"`
+    }
+  }
+  if (obj.assignment_id !== undefined) {
+    if (typeof obj.assignment_id !== 'string') {
+      return 'assignment_id must be a string'
+    }
+    if (obj.assignment_id.includes(' ')) {
+      return 'assignment_id must not contain spaces'
+    }
+  }
+  if (obj.tasks !== undefined) {
+    if (typeof obj.tasks !== 'object') {
+      return 'tasks must be an object'
+    }
+    // checks for each task
+    for (const taskKey in obj.tasks) {
+      if (taskKey.includes(' ')) {
+        return `Task key ${taskKey} must not contain spaces`
+      }
+      if (
+        !obj.tasks[taskKey].display_name ||
+        typeof obj.tasks[taskKey].display_name !== 'string'
+      ) {
+        return `Task ${taskKey} must have a display_name of type string`
+      }
+      if (
+        !obj.tasks[taskKey].short_display_name ||
+        typeof obj.tasks[taskKey].short_display_name !== 'string'
+      ) {
+        return `Task ${taskKey} must have a short_display_name of type string`
+      }
+      if (
+        obj.tasks[taskKey].blocking !== undefined &&
+        typeof obj.tasks[taskKey].blocking !== 'boolean'
+      ) {
+        return `Task ${taskKey} blocking must be a boolean`
+      }
+      if (
+        !obj.tasks[taskKey].color_hex ||
+        typeof obj.tasks[taskKey].color_hex !== 'string' ||
+        !hexColorRegex.test(obj.tasks[taskKey].color_hex)
+      ) {
+        return `Task ${taskKey} must have a valid color_hex (e.g. "#ff0000")`
+      }
+      if (
+        obj.tasks[taskKey].precondition === undefined ||
+        (obj.tasks[taskKey].precondition !== null &&
+          !(obj.tasks[taskKey].precondition in obj.tasks))
+      ) {
+        return `Task ${taskKey} precondition must be null or the key of another task`
+      }
+      for (const key in obj.tasks[taskKey]) {
+        if (!validTaskAttributes.includes(key)) {
+          return `Unknown attribute "${key}" in task "${taskKey}"`
+        }
+      }
+    }
+  }
+  if (obj.tasks !== undefined && obj.assignment_id === undefined) {
+    return 'Config also needs an assignment_id field if tasks are defined'
+  }
+  if (obj.assignment_id !== undefined && obj.tasks === undefined) {
+    return 'Config also needs a tasks field if assignment_id is defined'
+  }
+  for (const key in obj) {
+    if (!validAttributes.includes(key)) {
+      return `Unknown attribute "${key}"`
+    }
+  }
+  if (new TextEncoder().encode(JSON.stringify(obj)).length > MAX_JSON_SIZE) {
+    return 'The JSON object is too large. Maximum size is 10KB.'
+  }
+  //
+  // then validate the json with ajv (in case the above checks don't catch everything)
+  //
+  const ajv = new Ajv()
+  const schema = {
```

Comment:
Consider putting this in a separate file

Reply:
i tried to put the whole function into a seperate file ('cause it's big) but i get this error:
```
Module parse failed: Unexpected token (8:52) You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
```
Couldn't figure out how to configure the project to allow myself to put it in a separate file, but it's probably some small change tucked away somewhere